File: Challenge-2023-06/26-total-cost-to-hire-k-workers.go
// This is not correct, and I don't why why the priority queue does not work as expected

package main

import (
	"container/heap"
	"fmt"
)

type Item struct {
	value    int // The value of the item; arbitrary.
	priority int // The priority of the item in the queue.
	// The index is needed by update and is maintained by the heap.Interface methods.
	index int // The index of the item in the heap.
}

// A PriorityQueue implements heap.Interface and holds Items.
type PriorityQueue []*Item

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	// We want Pop to give us the highest, not lowest, priority so we use greater than here.
	return pq[i].priority < pq[j].priority
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x any) {
	n := len(*pq)
	item := x.(*Item)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() any {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil  // avoid memory leak
	item.index = -1 // for safety
	*pq = old[0 : n-1]
	return item
}

// update modifies the priority and value of an Item in the queue.
func (pq *PriorityQueue) update(item *Item, value int, priority int) {
	item.value = value
	item.priority = priority
	heap.Fix(pq, item.index)
}

func totalCost(costs []int, k int, candidates int) int64 {
	var res int64 = 0
	n := len(costs)
	left := candidates - 1
	right := n - candidates
	count := 0
	pq1 := make(PriorityQueue, candidates)
	pq2 := make(PriorityQueue, candidates)
	usedMap := make(map[int]bool)
	for i := 0; i < n; i++ {
		usedMap[i] = false
	}
	for i := 0; i < candidates; i++ {
		pq1[i] = &Item{
			value:    i,
			priority: costs[i],
			index:    i,
		}
		pq2[i] = &Item{
			value:    n - i - 1,
			priority: costs[n-i-1],
			index:    i,
		}
	}
	heap.Init(&pq1)
	heap.Init(&pq2)
	for left < n && right >= 0 && count < k {
		fmt.Println("pq1")
		for x := 0; x < pq1.Len(); x++ {
			fmt.Print(pq1[x].priority, ",")
		}
		fmt.Println()
		fmt.Println("pq2")
		for x := 0; x < pq2.Len(); x++ {
			fmt.Print(pq2[x].priority, ",")
		}
		fmt.Println()
		fmt.Println("count =", count, "|", left, ",", right)
		lMin := pq1[0]
		rMin := pq2[0]
		if lMin.priority < rMin.priority {
			res += int64(lMin.priority)
			usedMap[lMin.value] = true
			fmt.Println("selecting", lMin.priority, "from pq1")
			heap.Pop(&pq1)
			left++
			for left < n {
				if !usedMap[left] {
					pq1.Push(&Item{
						value:    left,
						priority: costs[left],
						index:    lMin.index,
					})
					break
				}
				left++
			}
		} else if lMin.priority > rMin.priority {
			res += int64(rMin.priority)
			usedMap[rMin.value] = true
			heap.Pop(&pq2)
			fmt.Println("selecting", rMin.priority, "from pq2")
			right--
			for right >= 0 {
				if !usedMap[right] {
					heap.Push(&pq2, &Item{
						value:    right,
						priority: costs[right],
						index:    rMin.index,
					})
					break
				}
			}
		} else {
			res += int64(lMin.priority)
			usedMap[lMin.value] = true
			pq1.Pop()
			fmt.Println("selecting", lMin.priority, "from pq1")
			left++
			for left < n {
				if !usedMap[left] {
					pq1.Push(&Item{
						value:    left,
						priority: costs[left],
						index:    lMin.index,
					})
					break
				}
				left++
			}
			// the value also on the right part
			if lMin.value >= right {
				fmt.Println("also selecting", rMin.priority, "from pq2")
				heap.Pop(&pq2)
				right--
				for right >= 0 {
					if !usedMap[right] {
						heap.Push(&pq2, &Item{
							value:    right,
							priority: costs[right],
							index:    rMin.index,
						})
						break
					}
					right--
				}
			}
		}
		count++
	}
	for count < k && pq1.Len() > 0 {
		res += int64(pq1[0].priority)
		fmt.Println(pq1[0].priority)
		heap.Pop(&pq1)
		count++
	}
	return res
}

func main() {
	costs := []int{28, 35, 21, 13, 21, 72, 35, 52, 74, 92, 25, 65, 77, 1, 73, 32, 43, 68, 8, 100, 84, 80, 14, 88, 42, 53, 98, 69, 64, 40, 60, 23, 99, 83, 5, 21, 76, 34}
	k := 32
	candidate := 12
	fmt.Println(totalCost(costs, k, candidate))
}

File: Challenge-2023-06/26-total-cost-to-hire-k-workers.cpp
#include "../header.h"

struct Item {
	int cost;
	int index;
	Item(int c, int i): cost{c}, index{i} {}
};

struct Compare {
	// sometimes, the top element in is not the one with least index if costs are same
	// so need to compare the index as the secondary condition
	bool operator()(Item i1, Item i2) {
		if (i1.cost > i2.cost) {
			return true;
		} else if (i1.cost < i2.cost) {
			return false;
		} else return i1.index > i2.index;
	}
};

class Solution {
public:
    long long totalCost(vector<int>& costs, int k, int candidates) {
        long long res = 0LL;
		int n = costs.size(), count = 0;
		int left = candidates - 1, right = n - candidates;
		unordered_set<int> usedSet;
		priority_queue<Item, vector<Item>, Compare> pq1, pq2;
		for (int i = 0; i < candidates; ++i) {
			pq1.push(Item(costs[i], i));
			pq2.push(Item(costs[n-i-1], n-i-1));
		}
		while (left < n && right >= 0 && count < k) {
			cout << "left = " << left << ", right = " << right << endl;
			Item lMin = pq1.top(), rMin = pq2.top();
			cout << "lMin.cost = " << lMin.cost << ", rMin.cost = " << rMin.cost << endl;
			cout << "lMin.index = " << lMin.index << ", rMin.index = " << rMin.index << endl;
			if (lMin.cost < rMin.cost) {
				res += (long long)lMin.cost;
				cout << "To add " << lMin.cost << endl;
				usedSet.insert(lMin.index);
				pq1.pop();
				left++;
				while (left < n) {
					if (usedSet.count(left) == 0) {
						pq1.push(Item(costs[left], left));
						break;
					}
					left++;
				}
			} else if (lMin.cost > rMin.cost) {
				res += (long long)rMin.cost;
				cout << "To add " << rMin.cost << endl;
				usedSet.insert(rMin.index);
				pq2.pop();
				right--;
				while (right >= 0) {
					if (usedSet.count(right) == 0) {
						pq2.push(Item(costs[right], right));
						break;
					}
					right--;
				}
			} else {
				cout << "To add " << lMin.cost << endl;
				if (lMin.index < rMin.index) {
					res += (long long)lMin.cost;
					usedSet.insert(lMin.index);
					pq1.pop();
					left++;
					while (left < n) {
						if (usedSet.count(left) == 0) {
							pq1.push(Item(costs[left], left));
							break;
						}
						left++;
					}
				} else if (rMin.index < lMin.index) {
					res += (long long)rMin.cost;
					usedSet.insert(rMin.index);
					pq2.pop();
					right--;
					while (right >= 0) {
						if (usedSet.count(right) == 0) {
							pq2.push(Item(costs[right], right));
							break;
						}
						right--;
					}
				} else {
					// a special case that left and right has overlapping
					cout << "\n ===== same value ===== \n" << endl;
					res += (long long)lMin.cost;
					usedSet.insert(lMin.index);
					pq1.pop();
					left++;
					while (left < n) {
						if (usedSet.count(left) == 0) {
							pq1.push(Item(costs[left], left));
							break;
						}
						left++;
					}
					pq2.pop();
					right--;
					while (right >= 0) {
						if (usedSet.count(right) == 0) {
							pq2.push(Item(costs[right], right));
							break;
						}
						right--;
					}
				}
			}
			count++;
			cout << endl << endl;
		}
		cout << "left = " << left << ", right = " << right << endl;
		cout << count << ", rest in pq1: " << pq1.size() << endl;
		while (count < k && !pq1.empty()) {
			Item lMin = pq1.top();
			res += (long long)lMin.cost;
			usedSet.insert(lMin.index);
			pq1.pop();
			count++;
		}
		return res;
    }
};

int main() {
	vector<int> costs {526,954,161,691,816,380,362,172,80,218,574,390,945,287,727,741,220,710,88,686,851,596,916,770,732,111,957,940,546,832,649,494,858,112,987,568,952,887,190,374,34,190,477,30,547,275,335,567,242,697,738,627,146,12,502,737,508,80,267,198,816,975,623,659,12,345,334,329,784,905,608,368,230,714,203,606,806,963,700,280,300,535,628,348,359,572,87,30,959,812,834};
	vector<int> orderCosts(costs.begin(), costs.end());
	sort(orderCosts.begin(), orderCosts.end());
	int res = 0;
	for (int i = 0; i < 72; ++i) {
		res += orderCosts[i];
	}
	cout << "Min costs = " << res << endl;
	int k = 72;
	int candidate = 48;
	cout << costs.size() << endl;
	cout << Solution().totalCost(costs, k, candidate) << endl;
	return 0;
}